Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,689 @@
+          /// occured and we should abort the user handler.
+          ///
+          /// Adopters are encouraged to cooperatively check for cancellation in their handlers but
+          /// we cannot rely on this.
+          ///
+          /// We additionally signal the handler that an error has occured by terminating the source
+          /// backing the request stream that the user handler is consuming.
+          ///
+          /// - NOTE: This handler has different semantics from the extant non-async-await handlers
+          /// where the `statusPromise` was explicitly failed with `GRPCStatus.unavailable` from
+          /// _outside_ the user handler. Here we terminate the request stream with a
+          /// `CancellationError` which manifests _inside_ the user handler when it tries to access
+          /// the next request in the stream. We have no control over the implementation of the user
+          /// handler. It may choose to handle this error or not. In the event that the handler
+          /// either rethrows or does not handle the error, this will be converted to a
+          /// `GRPCStatus.unknown` by `handleError(_:)`. Yielding a `CancellationError` _inside_
+          /// the user handler feels like the clearest semantics of what we want--"the RPC has an
+          /// error, cancel whatever you're doing." If we want to preserve the API of the
+          /// non-async-await handlers in this error flow we could add conformance to
+          /// `GRPCStatusTransformable` to `CancellationError`, but we still cannot control _how_
+          /// the user handler will handle the `CancellationError` which could even be swallowed.
+          ///
+          /// - NOTE: Currently we _have_ added `GRPCStatusTransformable` conformance to
+          /// `CancellationError` to convert it into `GRPCStatus.unavailable` and expect to
+          /// document that user handlers should always rethrow `CacellationError` if handled, after
+          /// optional cleanup.
+          requestStreamSource.finish(throwing: CancellationError())
+          responseStreamWriter._asyncWriter.cancelAsynchronously()
+        }
+        return .ok
+      }
+
+    case .active, .finishingSuccessfully:
+      self.handleError(GRPCError.ProtocolViolation("Multiple header blocks received on RPC"))
+
+    case .completed:
+      // We may receive headers from the interceptor pipeline if we have already finished (i.e. due
+      // to an error or otherwise) and an interceptor doing some async work later emitting headers.
+      // Dropping them is fine.
+      ()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMessage(_ request: Request) {
+    switch self.state {
+    case .idle:
+      self.handleError(GRPCError.ProtocolViolation("Message received before headers"))
+    case let .active(requestStreamSource, _, _, _):
+      // TODO: Handle YieldResult from `yield()`
+      requestStreamSource.yield(request)
```

Comment:
@glbrntt what should we be doing with the `YieldResult` here? Should we be checking that it is `.accepted`? And if it's `.dropped` do we consider this to be a fatal error that requires tearing down the RPC (because messages may have been lost)?

Reply:
OK, I've implemented these semantics and captured the reasoning in some commentary. I also dropped a `TODO` for the future intention to action the queue depth.